### ISSUE TITLE
Fuel-efficient routing makes small, unnecessary detour

### ISSUE THREAD

**Issue by mocsa** [osmandapp/OsmAnd#17177]
# 🐞 routing report

### Routing engine

- [x] OsmAnd's in-app offline routing
- [ ] Any online routing provider (YOURS, openrouteservice, OSRM, etc.)

### Routing Profile

Car with fuel-efficient way turned on.

### Start and end points

Start: https://osmand.net/map?pin=47.38866,19.11078#15/47.38866/19.11078

End: https://osmand.net/map?pin=47.39925,19.11569#16/47.39925/19.11569

### Actual and expected routes

The offline OsmAnd router makes a 10-15m unnecessary detour at [this point](https://osmand.net/map?pin=47.39166,19.11479#20/47.39166/19.11479). See screenshots. The detour disappears when fuel-efficient way is turned off.

I suspect that this is a problem in OsmAnd because I calculated the same route on openstreetmap.org, on graphhopper.com, and on openrouteservice.org and they did not make this detour.

The detour is also made on [test.osmand.net](https://test.osmand.net/map/?start=47.389895%2C19.113379&end=47.392713%2C19.114559&mode=car#18/47.39124/19.11499) when "Short way" is selected in "Additional Route Settings" (cogwheel icon next to the "Route profile" dropdown).

Birds eye view:

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/13969648/325df61e-d17b-4e8e-94da-e30da966db07"  width="300" height="auto"/> </th>
  </tr>
</table>  

Magnified view:

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/13969648/12f73767-8526-4492-b4ec-c2a22f1b928a"  width="300" height="auto"/> </th>
  </tr>
</table>  

Close view:

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/13969648/5a72ea0b-db31-4a6e-87b3-6e694dbd9f04"  width="300" height="auto"/> </th>
  </tr>
</table>  

### Is this a regression?

Don't know.

## 🌍 Your Environment

**OsmAnd Version:**

<pre><code>

OsmAnd~ 4.3.8
</code></pre>

**Device and Android/iOS version:**

Android 8.0.0

**Maps used (online or offline):**

- [x] Offline maps offered within the OsmAnd app for download.

- [ ] Online (tile / raster) maps 

Hungary map version 1st May 2023.

**Anything else relevant?**

**Comment by Zirochkabila** [comment]
@mocsa Please update the maps and OsmAnd - behavior isn't reproducible using the May maps.

**Comment by mocsa** [comment]
I updated to OsmAnd~ 4.4.7. The built-in downloader of OsmAnd says that the Hungary map version 1st May 2023 is the latest and that is what I'm using. The problem is reproducible with this updated setup, as well. The detour is made when fuel-efficient way is turned on.

The problem is also reproducible online. Please try the link to test.osmand.net in my issue above. Click on the link, then turn on the "Short way" option in Route profile. You will see the detour. I tried it a few minutes ago and it is there.

**Comment by Zirochkabila** [comment]
@mocsa You're right. The route does change when you activate "Short way", as you can see in the video - but, as you can see in the screenshots when you activate "Short way", the Cost value will also change, which is the conventional unit by which the route is evaluated.
If we are talking about this particular case, I think this is due to the presence of a traffic light and faster and "cheaper" routing occurs bypassing the traffic light.

<table style="width:100%">
  <tr>
    <th>"Short way" off</th>
    <th>"Short way" on</th>
  </tr>
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/104760013/bfba8bad-b9b6-4c9e-9530-57fa7bfb6b13"  width="500" height="auto"/> </th>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/104760013/0299f106-7ba1-4a77-9f35-76d1a5a90223"  width="500" height="auto"/> </th>
  </tr>
</table> 


https://github.com/osmandapp/OsmAnd/assets/104760013/9370d68f-e864-4b3a-bb50-463662546169

**Comment by mocsa** [comment]
I've decided to add a turn left restriction to the railway crossing node which is in the middle of this detour. Let me know if you have any better idea to correct this problem.

First I wanted to suggest that the routing algorithm should not suggest turns on a railway crossing node, but in this case 3 roads meet in such a node and the algorithm obviously cannot tell which road is actually crossing the railway and which are the connecting roads leading to the crossing.

I've also noticed that the lower (southmost) half of the detour was incorrect in OSM. It was tagged as a 2 lane road, having turn lanes. I've corrected it, but this did not fix the routing problem.
